test/local/devices_tests.js

```diff
@@ -0,0 +1,201 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+var uuid = require('uuid')
+var crypto = require('crypto')
+var test = require('../ptaptest')
+var mocks = require('../mocks')
+
+var modulePath = '../../lib/devices'
+
+test('require', function (t) {
+  t.equal(typeof require(modulePath), 'function', 'require returns function')
+  t.equal(require(modulePath).length, 3, 'returned function expects three arguments')
+
+  test('instantiate', function (t) {
+    var log = mocks.spyLog()
+    var deviceCreatedAt = Date.now()
+    var deviceId = crypto.randomBytes(16)
+    var device = {
+      name: 'foo',
+      type: 'bar'
+    }
+    var db = mocks.mockDB({
+      device: device,
+      deviceCreatedAt: deviceCreatedAt,
+      deviceId: deviceId
+    })
+    var push = mocks.mockPush()
+    var devices = require(modulePath)(log, db, push)
+
+    t.equal(typeof devices, 'object', 'devices is object')
+    t.equal(Object.keys(devices).length, 2, 'devices has two properties')
+
+    t.equal(typeof devices.upsert, 'function', 'devices has upsert method')
+    t.equal(devices.upsert.length, 3, 'devices.upsert expects three arguments')
+
+    t.equal(typeof devices.synthesizeName, 'function', 'devices has synthesizeName method')
+    t.equal(devices.synthesizeName.length, 1, 'devices.synthesizeName expects 1 argument')
```

Comment:
Reading between the lines, I guess what you're really asking is:

> What's the point of these assertions?

And maybe you have a point. Am I adding unnecessary maintenance burden?

Fwiw, my thinking is that unit tests should mock aggressively in order to run fast and expose their inner workings to assertions. And once you're mocking all the things, it makes sense to protect against unintentional interface (lower-case i) changes by asserting their shape and forcing people to update the tests if they're going to update the interface. Integration tests *might* catch those problems but also they might not, so it is better to be explicit about as much as possible in the unit tests. Plus, the unit tests give you quicker feedback while you're hacking.

But this is very much a home-grown philosophy and maybe the benefits I see / have seen in it aren't really there? Am re-considering it now! 😕